Python/Product/PythonTools/PythonTools/Editor/PythonTextBufferInfo.cs

```diff
@@ -473,12 +472,21 @@ sealed class PythonTextBufferInfo {
             int startCol = span.Start - span.Start.GetContainingLine().Start;
             int endCol = span.End - span.End.GetContainingLine().Start;
 
-            Tokenizer tokenizer = null;
 
+            // Cache may change from background thread during enumeration
+            // so make sure we track that snapshot version is still the same.
+            var spanSnapshot = span.Snapshot;
+
+            Tokenizer tokenizer = null;
             for (int line = firstLine; line <= lastLine; ++line) {
-                var lineTokenization = GetLineTokenization(span.Snapshot.GetLineFromLineNumber(line), ref tokenizer);
+                var currentBufferSpanshot = span.Snapshot.TextBuffer.CurrentSnapshot;
+                if (spanSnapshot.Version != currentBufferSpanshot.Version) {
+                    yield break;
```

Comment:
We probably need to abort more noisily here so that callers can restart with the new snapshot. At least auto-indent assumes that the end of iteration means we've reached the start of the file (and known state) - that will no longer be valid with this.

Reply:
Then again, it shouldn't actually matter here, as we do all the remapping required via tracking spans. But perhaps anyone iterating over these tokens will want to abort anyway if the content changes?

Reply:
It is probably better to lock the cache for the duration as it was doing before

Reply:
Agreed, but that will probably have to be done with LockedEnumerable and make sure that all callers properly dispose the enumerator (I think they're all using `foreach` so it should be fine).